### PR TITLE
Beats DRA pipeline

### PR DESCRIPTION
**Pull request by pazone**
### Beats Packaging pipeline
Migrates [packaging.groovy](https://github.com/elastic/beats/blob/main/.ci/packaging.groovy) to buildkite.
[Buildkite packaging pipeline](https://buildkite.com/elastic/beats-packaging-pipeline) is used to build and publish the packages for the Beats. The pipeline is triggered by a commit to the `main` or release branches.
The pipeline definition is located in the `.buildkite/packaging.pipeline.yml`

### Triggers
Staging packaging DRA is triggered for the `main` and release branches.
Snapshot can be triggered for any branch by the `/package` comment in the PR. The release-manager dry-run will be used for PR builds.

#### Beats dashboards

Generates `build/distributions/dependencies.csv` and adds it to the `beats-dashboards` artifact. The `dependencies.csv` is required by the release-manager configuration

#### Packaging snapshot/staging

- Builds the Beats packages for all supported platforms and architectures (`mage package, mage ironbank`)
- Copies artifacts `build/distributions/<beat>/` directory and adds it as an artifact. Where `<beat>` is the name of the beat
- x-pack artifacts a also copied to `build/distributions/<beat>/` directory, where `<beat>` is the name of the beat. For example, `auditbeat`, not `x-pack/auditbeat`

#### DRA publish
Downloads the artifacts from the `packaging snapshot/staging` step and publishes them to the Elastic DRA registry.

[Snapshot publishing example](https://buildkite.com/elastic/beats-packaging-pipeline/builds/23) (Dry run)

Beats-tester and e2e tests are not triggered

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [x] All Snapshot and staging artifacts are built, present, and properly transferred to the corresponding DRA steps
- [x] DRA dry run passes 
- [ ] All artifacts have the same size and permissions as on Jenkins. (Need to check one more time)

## Related issues

- https://github.com/elastic/ingest-dev/issues/3072
- https://github.com/elastic/ingest-dev/issues/3095

### CODE REVIEW

**Comment by pazone** [comment]
Should I disable the Jenkins package pipeline before we merge that?

**Comment by pierrehilbert** [comment]
> Should I disable the Jenkins package pipeline before we merge that?

From what I understood yesterday during our sync, we shouldn't have both running in the same time to avoid conflicts so I would say yes but will let @dliappis confirms here.

File: .buildkite/scripts/packaging/package-dra.sh
Comment (alexsapran) [inline .buildkite/scripts/packaging/package-dra.sh:1]: We are missing the `set -ueo pipefail` so that if a step fails the script fails

File: .buildkite/scripts/packaging/package-dra.sh
Comment (alexsapran) [inline .buildkite/scripts/packaging/package-dra.sh:2]: `${1:-""}` so that if `-u` will not complain

**Comment by dliappis** [comment]
@pav

> > Should I disable the Jenkins package pipeline before we merge that?
> 
> From what I understood yesterday during our sync, we shouldn't have both running in the same time to avoid conflicts so I would say yes but will let @dliappis confirms here.

I replied in https://github.com/elastic/ingest-dev/issues/3095#issuecomment-2074220950, let's move the discussion about the mechanics of testing there.

File: catalog-info.yaml
Comment (dliappis) [inline catalog-info.yaml:1058]: This is a duplicate of the one already present (in `main`): https://github.com/elastic/beats/blob/3ba900be92cf4c207787f05a08c6d5d5fcc8fcd2/catalog-info.yaml#L1020-L1054 so let's remove it?

**Review by dliappis — CHANGES_REQUESTED** [review]
Did a 5 min pass to familiarize myself with the high level direction.
Will continue the review a bit later, for now left a comment for a duplicate section in catalog-info.yaml

File: .buildkite/pull-requests.json
Comment (v1v) [inline .buildkite/pull-requests.json:61]: What's the reason for these comments if it's not listening for PRs (or that's what I see in the `catalog-info.yaml`)?

File: .buildkite/packaging.pipeline.yml
Comment (v1v) [inline .buildkite/packaging.pipeline.yml:50]: Do you think a top-level env variable with the different combinations could help?

```
env:
  PLATFORMS_AMD: " "
  PLATFORMS_ARM: " "

...

- label: "..."
  env:
    PLATFORMS: "${PLATFORMS_AMD}"
...
```

File: .buildkite/scripts/packaging/package-dra.sh
Comment (alexsapran) [inline .buildkite/scripts/packaging/package-dra.sh:9]: What about prefixing with somethink like `~~~` or `+++` depending on if we want it to be expanded or not in BK ui.

File: .buildkite/packaging.pipeline.yml
Comment (v1v) [inline .buildkite/packaging.pipeline.yml:59]: Do you think you can use [anchors](https://medium.com/@kinghuang/docker-compose-anchors-aliases-extensions-a1e4105d70bd) to help with the list of supported beats in the matrix?

File: .buildkite/pull-requests.json
Comment (alexsapran) [inline .buildkite/pull-requests.json:51]: do we need this? are we going to perform this operation from a PR basis?

File: .buildkite/packaging.pipeline.md
Comment (v1v) [inline .buildkite/packaging.pipeline.md:7]: IIUC, `PRs` are not supported, or I cannot see those changes in the `catalog-info.yaml`.

File: .buildkite/packaging.pipeline.md
Comment (v1v) [inline .buildkite/packaging.pipeline.md:2]: The sections in `triggers` said about PRs but that's not reflected here.

File: catalog-info.yaml
Comment (alexsapran) [reply catalog-info.yaml:1058]: I was about to comment something similar, how you were able to test the pipeline while the pipeline itself is not yet merged (thus created). 
From the looks of it, you the above pipeline and this, share the same `pipeline_file` so I assume adding a new backstage entity is something we dont want to do.

File: .buildkite/packaging.pipeline.md
Comment (v1v) [inline .buildkite/packaging.pipeline.md:19]: I wonder whether a link to what's the rationale for this, maybe the Unified Release team has some details about this?

**Review by alexsapran — COMMENTED** [review]
did a first review

File: .buildkite/scripts/dra.sh
Comment (v1v) [inline .buildkite/scripts/dra.sh:1]: Missed the shebang
```suggestion
#!/usr/bin/env bash
set -uo pipefail
```

File: .buildkite/scripts/dra.sh
Comment (v1v) [inline .buildkite/scripts/dra.sh:6]:
```suggestion
if [[ "${BUILDKITE_PULL_REQUEST:="false"}" != "false" || "${BUILDKITE_BRANCH}" == "ci_packaging_pipeline" ]]; then
```

File: .buildkite/packaging.pipeline.yml
Comment (v1v) [inline .buildkite/packaging.pipeline.yml:211]: **super-nit**: what's the reason for having all those steps in the pipeline rather than a centralised script called `dra-publish.sh` that calls all of them?

File: .buildkite/packaging.pipeline.yml
Comment (alexsapran) [inline .buildkite/packaging.pipeline.yml:217]: Staging should be only release branches, so the condition should be 
```yaml
        if: "build.branch =~ /\\d+\\.\\d+/
```
To make this a less strict condition we could have something like
```yaml
        if: "build.branch =~ /\\d+\\.\\d+/ || build.env("RUN_STAGING") == true
```

The name ofc is subject for discussions.

File: .buildkite/packaging.pipeline.yml
Comment (alexsapran) [inline .buildkite/packaging.pipeline.yml:203]: the snapshot is something that we are running for the release branch and the main branch, so this step should have the following condition
```yaml
if: "build.branch =~ /\\d+\\.\\d+/ || build.branch == 'main'"
```

Additionally, having an env var to loose the strictness, similar to the staging step bellow, would be good.

File: .buildkite/scripts/dra.sh
Comment (dliappis) [inline .buildkite/scripts/dra.sh:37]:
```suggestion
  DRA_CREDS_SECRET=$(retry -t 5 -- vault kv get -field=data -format=json ${CI_DRA_ROLE_PATH})
```

File: .buildkite/scripts/dra.sh
Comment (dliappis) [inline .buildkite/scripts/dra.sh:17]: you can remove all this function once you've switched `retry 5` to `retry -t 5 --` as shown in the ticket below

File: .buildkite/packaging.pipeline.yml
Comment (dliappis) [reply .buildkite/packaging.pipeline.yml:50]: Btw this is probably to the approach taken for migrated pipelines to be explicit when the variable is referenced in just a few steps e.g. https://github.com/elastic/beats/blob/main/.buildkite/x-pack/pipeline.xpack.heartbeat.yml#L231

If, we found ourselves in this pipeline that we need to copy/paste it in most places, agreed, it makes sense to reduce the copy/pasting.

File: .buildkite/pull-requests.json
Comment (pazone) [reply .buildkite/pull-requests.json:61]: That was done to make it exactly as it was made on Jenkins. But I see that no one likes this functionality. Removing it

Comment (v1v) [reply .buildkite/pull-requests.json:61]: I didn't say to remove it, I remarked that the `catalog-info.yaml` is not configured to listen for PRs.

IMO, `/package` is a valuable GitHub command, but that's only my persona view.

File: catalog-info.yaml
Comment (pazone) [reply catalog-info.yaml]: Done

File: .buildkite/pull-requests.json
Comment (pazone) [reply .buildkite/pull-requests.json]: I can add it in later phases. It now uses 30 VMs. The less we trigger it - the better

File: .buildkite/packaging.pipeline.md
Comment (pazone) [reply .buildkite/packaging.pipeline.md]: Added the link

File: .buildkite/packaging.pipeline.yml
Comment (pazone) [reply .buildkite/packaging.pipeline.yml:207]: Updated conditions

File: .buildkite/scripts/dra.sh
Comment (pazone) [reply .buildkite/scripts/dra.sh]: Done

File: .buildkite/packaging.pipeline.yml
Comment (pazone) [reply .buildkite/packaging.pipeline.yml]: Done. Please take a look again

File: .buildkite/scripts/dra.sh
Comment (pazone) [reply .buildkite/scripts/dra.sh:2]: 👍

File: .buildkite/packaging.pipeline.md
Comment (dliappis) [inline .buildkite/packaging.pipeline.md:6]:
```suggestion
Staging packaging DRA is triggered for release branches only.
```

File: .buildkite/packaging.pipeline.yml
Comment (pazone) [reply .buildkite/packaging.pipeline.yml:215]: IMHO it makes it a bit easier to understand what steps we perform during the release.

File: .buildkite/packaging.pipeline.md
Comment (dliappis) [inline .buildkite/packaging.pipeline.md:7]:
```suggestion
Snapshot is triggered for `main` and release branches.
```

File: .buildkite/packaging.pipeline.md
Comment (dliappis) [inline .buildkite/packaging.pipeline.md:12]:
```suggestion
Generates `build/distributions/dependencies.csv` and `tar.gz` and adds them to the `beats-dashboards` artifact. This is required by the release-manager configuration.
```

File: .buildkite/scripts/packaging/package-dra.sh
Comment (pazone) [reply .buildkite/scripts/packaging/package-dra.sh:1]: Done

**Review by dliappis — CHANGES_REQUESTED** [review]
This is very close to LGTM from my side!

The main ask is to remove the specific sections due to the lack of msitools (we have it now) and to check again the addition in pull-requests.json

File: .buildkite/pull-requests.json
Comment (pazone) [reply .buildkite/pull-requests.json:53]: oops

**Comment by pazone** [comment]
@dliappis Removed `install-msitools` script invocation. Testing everything again

File: .buildkite/pull-requests.json
Comment (alexsapran) [inline .buildkite/pull-requests.json:37]: https://github.com/elastic/beats/pull/39154 is merged, how this is showing on the PR?

File: .buildkite/packaging.pipeline.yml
Comment (pazone) [reply .buildkite/packaging.pipeline.yml]: Extracted the variables

File: .buildkite/packaging.pipeline.yml
Comment (pazone) [reply .buildkite/packaging.pipeline.yml]: Done

File: .buildkite/pull-requests.json
Comment (pazone) [reply .buildkite/pull-requests.json]: Reverted the changes

File: .buildkite/pull-requests.json
Comment (pazone) [reply .buildkite/pull-requests.json:37]: I think I need to rebase my branch. But after that we won't be able to test packaging

File: .buildkite/packaging.pipeline.yml
Comment (pazone) [reply .buildkite/packaging.pipeline.yml]: Left an exclusion for agentbeat because it need a bigger vm

File: .buildkite/scripts/dra.sh
Comment (dliappis) [inline .buildkite/scripts/dra.sh:5]: I recommend we change it to

```
if [[ "$DRY_RUN" == "false" ]]; then echo "--- Running in publish mode"; DRY_RUN=""; else echo "--- Running in dry-run mode"; DRY_RUN="--dry-run"; fi
```

to facilitate easy testing.

**Review by dliappis — APPROVED** [review]
LGTM!

File: .buildkite/scripts/packaging/prepare-release-manager.sh
Comment (v1v) [inline .buildkite/scripts/packaging/prepare-release-manager.sh:1]: As far as I see this file is a copy of `.ci/scripts/prepare-release-manager.sh` but `.ci/scripts/prepare-release-manager.sh` has not been deleted...

**Pull request merged**
